### ISSUE TITLE
Generalized PPO loss

### ISSUE THREAD

**Issue by qingquansong** [linkedin/Liger-Kernel#626]
### 🚀 The feature, motivation and pitch

- [ ] 1. Current GRPO assume KL term is added and advantage is computed in side the loss, we wanna open this to become configurable from user
- [ ] 2. Clipping is not used in GRPO and need to be added (also as an option for generalized PPO case)
- [ ] 3. Reference model logits/prob can be provided directly without providing LM head + hidden states as an option
- [ ] 4. Old policy prob or LM head/hidden states is not provided for importance sampling purpose in GRPO.

**Comment by qingquansong** [comment]
Assigned to me now and if anyone is interested in anyone of the features, feel free to take.

**Comment by kashif** [comment]
yes the implementation in liger is older so would need updating to the newer version in TRL

**Comment by shivam15s** [comment]
worked on #628 ... didn't see this issue before so was working on this independently 
 1. KL is only calculated if beta is non-zero. Assume user directly provides the advantage instead of reward (otherwise we need all gather to compute reward mean/std)
 2. Added clipping
 3. Not supported yet 
 4. added an option for old policy prob
Also added fix for loss correctness, metrics calculation, testing, etc.
